flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java

```diff
@@ -357,15 +357,16 @@ public static TaskExecutor startTaskManager(
 		TaskManagerServices taskManagerServices = TaskManagerServices.fromConfiguration(
 			taskManagerServicesConfiguration,
 			resourceID,
+			remoteAddress,
+			metricRegistry,
```

Comment:
Why do we need to move `taskManagerMetricGroup` from `TaskExecutor` to `TaskManagerServices`? It seems it would be enough to just create `taskManagerMetricGroup` before `taskManagerServices` and pass it to `TaskManagerServices.fromConfiguration` and have less changes.

Also, I am wondering whether it would make sense to pull out the creation of `NetworkEnviroment` completely from `TaskManagerServicesConfiguration.fromConfiguration`/`TaskManagerServices.fromConfiguration` and pass it as a parameter to `TaskManagerServices.fromConfiguration`. We could migrate completely `NetworkEnviroment` creation into `NetworkEnviroment.fromConfiguration(configuration)` to get closer to `ShuffleMaster.createShuffleService(configuration)`. Probably, as a separate commit.

Reply:
Yes it also makes sense to create the `TaskManagerMetricGroup` before `TaskManagerServices`. I just thought it should be regarded as a component of `TaskManagerServices` like `TaskManagerLocation`. Nevertheless from the aspect of minimal change, I am willing to make it separate as you suggested. 

That is a good idea to extract the creation of `NetworkEnvironment` completely. I would do that in separate commit.

Reply:
Considering the issue of `NetworkEnviroment.fromConfiguration(configuration)`, we should decouple the `NetworkEnvironmentConfiguration` from `TaskManagerServicesConfiguration` and the current creation of `NetworkEnvironmentConfiguration` also relies on other parameters such as `maxJvmHeapMemory`, `remoteAddress` and `localCommunication`.

So it might be involved in more refactoring to create `NetworkEnvironment` from flink `Configuration` directly. Then I want to create a separate ticket for this issue. What do you think?

Reply:
agreed 👍